### DISCUSSION TITLE
ros_deploy_base docker container rule

### DISCUSSION THREAD

**Discussion by thomasegriffith** [mvukov/rules_ros2#98] [General]
Hey,

I'm seeing a couple peculiarities related to the `ros_deploy_base` docker container rule:

1. When running the `//chatter:chatter_image` example from the `examples` folder, I run into the below error:

```
$ bazel run //chatter:chatter_image
Starting local Bazel server and connecting to it...
INFO: Analyzed target //chatter:chatter_image (166 packages loaded, 21057 targets configured).
INFO: Found 1 target...
Target //chatter:chatter_image up-to-date:
  bazel-bin/chatter/chatter_image-layer.tar
INFO: Elapsed time: 90.353s, Critical Path: 60.65s
INFO: 869 processes: 200 internal, 669 linux-sandbox.
INFO: Build completed successfully, 869 total actions
INFO: Running command line: bazel-bin/chatter/chatter_image.executable
92bda4a116f1: Loading layer [==================================================>]  192.8MB/192.8MB
Loaded image ID: sha256:729e23d3d84dd21ec9aa822b884820eeb7d4b2629a099fb32561a16783ac5a8b
Tagging 729e23d3d84dd21ec9aa822b884820eeb7d4b2629a099fb32561a16783ac5a8b as bazel/chatter:chatter_image
[INFO] [launch]: All log files can be found below /root/.ros/log/2023-04-17-22-05-48-997969-tommy-GE76-Raider-11UE-1
[INFO] [launch]: Default logging verbosity is set to INFO
[ERROR] [launch]: Caught exception in launch (see debug for traceback): Caught exception when trying to load file of format [py]: libatomic.so.1: cannot open shared object file: No such file or directory
The C extension '/app/chatter/chatter.runfiles/ros2_rclpy/rclpy/rclpy/_rclpy_pybind11.so' failed to be imported while being present on the system. Please refer to 'https://docs.ros.org/en/{distro}/Guides/Installation-Troubleshooting.html#import-failing-even-with-library-present-on-the-system' for possible solutions
```

2.) I'm also trying to use `@ros_deploy_base//image` as a base image to a `rules_docker` build target that uses a `ros2_py_binary` build target (that depends on `rosbag2`) as the binary:

```
ros2_py_binary (
    name = "bar",
    srcs = ["bar.py"],
    main = "bar.py",
    set_up_ament = True,
    deps = [
        "@ros2_rosbag2//:rosbag2_py",
    ],
)

container_image(
    name = "bar_image",
    base = "@ros_deploy_base//image",
    binary = ":bar",
    docker_run_flags = "-it --rm --network=host",
)
```

When running `bazel run` on it, I get the following error:

```
  File "/app/foo/bar.runfiles/ros2_rosbag2/rosbag2_py/rosbag2_py/__init__.py", line 21, in <module>
    from rosbag2_py._reader import (
ImportError: /lib/x86_64-linux-gnu/libstdc++.so.6: version `CXXABI_1.3.13' not found (required by /app/foo/bar.runfiles/ros2_rosbag2/rosbag2_py/rosbag2_py/_reader.so)
```

I can get around the problem in my own use-case by just using a base image that has ROS2 installed.  This does require a ROS2 install, which isn't the end of the world, but is perhaps not ideal since this should probably work without a ROS2 install.

---

I saw that `ros_deploy_base` hasn't been updated in a couple years.

I'm wondering if `ros_deploy_base` should be updated?  Would that help with my problems?  Or is there perhaps something else at work here?

Thanks!

**Comment by mvukov** [comment]
The original intent with `ros_deploy_base` was to demonstrate a very simple use-case. In your first use-case, I think you need to install `libatomic1` via apt. The second issue is maybe related to OS incompatibility: e.g. binary compiled on one OS and the base image on Ubuntu 20.04, as demonstrated in https://github.com/mvukov/rules_ros/blob/main/docker/base/base.Dockerfile#L1. A good practice IMO is to keep the dev/CI container (where one compiles code) and the base container for deployment based on the same OS container, e.g. `ubuntu:20.04`. Different users will have different requirements, so, I suggest you create a base image that fits your needs.

**Comment by thomasegriffith** [comment]
Sorry for the late reply.  That makes sense.  Thank you!
